### PR TITLE
Implemented interface to define new types of Kalman filters

### PR DESCRIPTION
**Pull request by guilhermebodin**

### CODE REVIEW

File: src/StateSpaceModels.jl [hunk 1]
```diff
@@ -25,24 +26,22 @@ function statespace(model::StateSpaceModel; nseeds::Int = 3, verbose::Int = 1)
     end
 
     # Maximum likelihood estimation
-    param = estimate_statespace(model, nseeds; verbose = verbose)
+    covariance = estimate_statespace(model, model.optimization_method; verbose = verbose)
 
     if verbose > 0
         @info("End of estimation.")
-        @info("Starting filtering and smoothing...")
```
Comment (raphaelsaavedra) [inline src/StateSpaceModels.jl]: Any reason why the infos were removed? I think there should be at least one info talking about the filtering/smoothing if there is one for estimating.

File: src/sqrt_kalman.jl [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        sqrtN[:, :, t-1] = NstarG[1:m, 1:m]
+
+        # Smoothed state and its covariance
+        P = gram(sqrtP[:, :, t])
+        N = gram(sqrtN[:, :, t])
+        alpha[t, :] = a[t, :] + P * r[t-1, :]
+        V[:, :, t]  = P - (P * N * P)
+    end
+
+    F_1 = gram(sqrtF[:, :, 1])
```
Comment (raphaelsaavedra) [inline src/sqrt_kalman.jl]: This variable is only used once, no need to save it

File: src/StateSpaceModels.jl (hunk 1 above)
Comment (guilhermebodin) [reply src/StateSpaceModels.jl]: ok

File: src/sqrt_kalman.jl (hunk 2 above)
Comment (guilhermebodin) [reply src/sqrt_kalman.jl]: done

File: src/sqrt_kalman.jl [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    L[:, :, 1]   = T - U2star[:, :, 1] * pinv(sqrtF[:, :, 1]) * Z[:, :, 1]
```
[the remark this answers is not in the document]
Comment (guilhermebodin) [reply src/sqrt_kalman.jl]: done

File: src/sqrt_kalman.jl [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    r_0    = Z[:, :, 1]' * pinv(F_1) * v[1, :] + L[:, :, 1]' * r[1, :]
+    Nstar  = [Z[:, :, 1]' * pinv(sqrtF[:, :, 1]) L[:, :, 1]' * sqrtN[:, :, 1]]
+    G      = qr(Nstar').Q
+    NstarG = Nstar*G
```
[the remark this answers is not in the document]
Comment (guilhermebodin) [reply src/sqrt_kalman.jl]: done

File: src/structures.jl [hunk 5]
```diff
@@ -1,5 +1,61 @@
...
+"""
+mutable struct RandomSeedsLBFGS <: AbstractOptimizationMethod
+    f_tol::Float64
+    g_tol::Float64
+    iterations::Int
+    nseeds::Int
+    seeds::Array{Float64}
+
+    function RandomSeedsLBFGS()
+        return new(1e-10, 1e-10, 1e5, 3)
```
[the remark this answers is not in the document]
Comment (guilhermebodin) [reply src/structures.jl]: changed to 1e-6

**Pull request merged**